File: 40.combination-sum-ii.py
#
# @lc app=leetcode id=40 lang=python3
#
# [40] Combination Sum II
#
from typing import *
# @lc code=start


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        width = len(candidates)
        pos_vec = [0]
        sum = 0
        res = []
        while True:
            if pos_vec[-1] != width:
                if sum+candidates[pos_vec[-1]] == target:
                    res.append([candidates[i] for i in pos_vec])
                elif sum+candidates[pos_vec[-1]] < target:
                    sum += candidates[pos_vec[-1]]
                    pos_vec.append(pos_vec[-1]+1)
                    continue
            if pos_vec[-1] == width or sum+candidates[pos_vec[-1]] >= target:
                pos_vec.pop()
                if len(pos_vec) == 0:
                    break
                sum -= candidates[pos_vec[-1]]
                pos_vec[-1] += 1
                while pos_vec[-1] != width and candidates[pos_vec[-1]] == candidates[pos_vec[-1]-1]:
                    pos_vec[-1] += 1
        return res
# @lc code=end

File: 40.combination-sum-ii.cpp
/*
 * @lc app=leetcode id=40 lang=cpp
 *
 * [40] Combination Sum II
 */
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;
// @lc code=start
class Solution
{
public:
    vector<vector<int>> combinationSum2(vector<int> &candidates, int target)
    {
        sort(candidates.begin(), candidates.end());
        int width = candidates.size();
        vector<int> pos_vec = {0};
        int sum = 0;
        vector<vector<int>> res;
        while (true)
        {
            if (pos_vec.back() != width)
            {
                if (sum + candidates[pos_vec.back()] == target)
                {
                    res.push_back({});
                    for (auto pos : pos_vec)
                    {
                        res.back().push_back(candidates[pos]);
                    }
                }
                else if (sum + candidates[pos_vec.back()] < target)
                {
                    sum += candidates[pos_vec.back()];
                    pos_vec.push_back(pos_vec.back() + 1);
                    continue;
                }
            }
            if (pos_vec.back() == width || sum + candidates[pos_vec.back()] >= target)
            {
                pos_vec.pop_back();
                if (pos_vec.empty())
                {
                    break;
                }
                sum -= candidates[pos_vec.back()];
                ++pos_vec.back();
                while (pos_vec.back() != width && candidates[pos_vec.back()] == candidates[pos_vec.back() - 1])
                {
                    pos_vec.back() += 1;
                }
            }
        }
        return res;
    }
};
// @lc code=end
